utils/resb/src/binary/deserializer.rs

```diff
@@ -0,0 +1,1117 @@
+    where
+        V: de::Visitor<'de>,
+    {
+        visitor.visit_u32(self.parse_unsigned()?)
+    }
+
+    fn deserialize_u64<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        visitor.visit_u64(self.parse_unsigned()? as u64)
+    }
+
+    fn deserialize_f32<V>(self, _visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        // Resource bundles have no native concept of floating point numbers and
+        // no examples of storing them have been encountered.
+        unimplemented!()
+    }
+
+    fn deserialize_f64<V>(self, _visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        // Resource bundles have no native concept of floating point numbers and
+        // no examples of storing them have been encountered.
+        unimplemented!()
+    }
+
+    fn deserialize_char<V>(self, _visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        // Resource bundles have no native concept of single characters and no
+        // examples of storing them have been encountered.
+        unimplemented!()
+    }
+
+    fn deserialize_str<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        // Strings in resource bundles are stored as UTF-16 and can't be
+        // borrowed.
+        self.deserialize_string(visitor)
+    }
+
+    fn deserialize_string<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        let descriptor = self.get_next_resource_descriptor()?;
+        match descriptor.resource_type() {
+            ResourceReprType::_String => todo!(),
+            ResourceReprType::StringV2 => {
+                if let Some(data_16_bit) = self.data_16_bit {
+                    if descriptor.is_empty() {
+                        return visitor.visit_string(String::from(""));
+                    }
+
+                    let input = get_subslice(
+                        data_16_bit,
+                        get_16_bit_offset(descriptor.offset() as usize)..,
+                    )?;
+                    let de = Resource16BitDeserializer::new(input);
+                    de.deserialize_string(visitor)
+                } else {
+                    Err(Error::new(
+                        "Invalid resource bundle: StringV2 resource without 16-bit data block.",
+                    ))
+                }
+            }
+            _ => Err(Error::new(
+                "Unexpected resource type where string expected.",
+            )),
+        }
+    }
+
+    fn deserialize_bytes<V>(self, visitor: V) -> Result<V::Value, Self::Error>
+    where
+        V: de::Visitor<'de>,
+    {
+        let descriptor = self.get_next_resource_descriptor()?;
+        let value = match descriptor.resource_type() {
```

Comment:
Thought: This branching seems a little bit heavy, in the sense that we need to retain all three branches even though we know ahead of time which branch to expect. It would be nice to be able to say, "I expect an IntVector here; please give it to me as bytes." But I guess this is a limitation of the Serde data model.

Reply:
Yeah, we don't really have any way of hinting to the deserializer that we know what's coming up, unfortunately.

Reply:
Thought: For the future, we could add functions outside of the `Serde::Deserializer` trait to this struct, like `pub fn deserialize_string_v2_as_bytes`, which might be useful for code size or efficiency in hot code paths.